src/v/cluster/shard_placement_table.h

```diff
@@ -0,0 +1,172 @@
+
+        // invariant: if local is nullopt, then expected_on_this_shard() is true
+        // (there is no point in keeping an item that has no local state as well
+        // as is not expected on this shard)
+        bool expected_on_this_shard() const {
+            return !_next && target && target->shard == ss::this_shard_id();
+        }
+
+        friend std::ostream& operator<<(std::ostream&, const placement_state&);
+
+    private:
+        friend class shard_placement_table;
+
+        placement_state() = default;
+        placement_state(
+          const shard_placement_target& target, model::shard_revision_id rev)
+          : target(target)
+          , shard_revision(rev) {}
+
+        /// If this shard is the initial shard for this partition on this node,
+        /// this field will contain the corresponding shard revision.
+        model::shard_revision_id _is_initial_at_revision;
+        /// If x-shard transfer is in progress, will hold the destination.
+        std::optional<ss::shard_id> _next;
+    };
+
+    // must be called on each shard
+    ss::future<> initialize(const topic_table&, model::node_id self);
+
+    using shard_callback_t
+      = std::function<void(const model::ntp&, model::shard_revision_id)>;
+
+    /// Must be called only on assignment_shard_id. Shard callback will be
+    /// called on shards that have been modified by this invocation.
+    ss::future<> set_target(
+      const model::ntp&,
+      std::optional<shard_placement_target>,
+      model::shard_revision_id,
+      shard_callback_t);
+
+    // getters
+
+    std::optional<placement_state> state_on_this_shard(const model::ntp&) const;
+
+    // partition lifecycle methods
+
+    ss::future<std::error_code>
+    prepare_create(const model::ntp&, model::revision_id expected_log_rev);
+
+    // return value is a tri-state:
+    // * if it returns a shard_id value, a transfer to that shard must be
+    // performed
+    // * if it returns errc::success, transfer has already been performed
+    // * else, we must wait before we begin the transfer.
+    ss::future<result<ss::shard_id>>
+    prepare_transfer(const model::ntp&, model::revision_id expected_log_rev);
+
+    ss::future<> finish_transfer_on_destination(
+      const model::ntp&, model::revision_id expected_log_rev);
+
+    ss::future<> finish_transfer_on_source(
+      const model::ntp&, model::revision_id expected_log_rev);
+
+    ss::future<std::error_code>
+    finish_delete(const model::ntp&, model::revision_id expected_log_rev);
+
+private:
+    void set_target_on_this_shard(
+      const model::ntp&,
+      std::optional<shard_placement_target>,
+      model::shard_revision_id,
+      bool is_initial,
+      shard_callback_t);
+
+    ss::future<> do_delete(const model::ntp&, placement_state&);
+
+private:
+    // per-shard state
+    //
+    // node_hash_map for pointer stability
+    absl::node_hash_map<model::ntp, placement_state> _states;
```

Comment:
consider using a hierarchical data structure to save memory

Reply:
I am actually on a fence about this as it would complicate the code considerably (we are constantly adding and removing stuff to/from this map). And we need it to be per-ntp (not per-group for example), as we have to be able to find previous incarnation of the partition with the same ntp to delete it before creating the new instance.